File: test-suite-python/src/main/python/helloworld/CustomInterceptor.py
# tag::imports[]
from io.grpc import Metadata, ServerCall, ServerCallHandler, ServerInterceptor
from jakarta.inject import Singleton
from micronaut.core.order import Ordered
# end::imports[]
from micronaut.context.annotation import Requires

INTERCEPTED: list[str] = []


@Requires(property="spec.name", value="ServerInterceptorTest")
# tag::clazz[]
@Singleton  # <1>
class CustomInterceptor(ServerInterceptor, Ordered):  # <2>

    def interceptCall[T, R](
        self,
        call: ServerCall[T, R],
        headers: Metadata,
        next: ServerCallHandler[T, R],
    ) -> ServerCall.Listener[T]:
        # end::clazz[]
        INTERCEPTED.append(call.getMethodDescriptor().getFullMethodName())
        # tag::clazz[]
        return next.startCall(call, headers)

    def getOrder(self) -> int:
        return 10  # <3>
# end::clazz[]

File: test-suite/src/main/java/helloworld/CustomInterceptor.java
package helloworld;

// tag::imports[]
import io.grpc.Metadata;
import io.grpc.ServerCall;
import io.grpc.ServerCallHandler;
import io.grpc.ServerInterceptor;
import io.micronaut.core.order.Ordered;
import jakarta.inject.Singleton;
// end::imports[]
import io.micronaut.context.annotation.Requires;

import java.util.List;
import java.util.concurrent.CopyOnWriteArrayList;

@Requires(property = "spec.name", value = "ServerInterceptorTest")
// tag::clazz[]
@Singleton // <1>
public class CustomInterceptor implements ServerInterceptor, Ordered { // <2>
    // end::clazz[]

    static final List<String> INTERCEPTED = new CopyOnWriteArrayList<>();

    // tag::clazz[]
    @Override
    public <T, R> ServerCall.Listener<T> interceptCall(ServerCall<T, R> call,
                                                       Metadata headers,
                                                       ServerCallHandler<T, R> next) {
        // end::clazz[]
        INTERCEPTED.add(call.getMethodDescriptor().getFullMethodName());
        // tag::clazz[]
        return next.startCall(call, headers);
    }

    @Override
    public int getOrder() {
        return 10; // <3>
    }
}
// end::clazz[]
